ios-training-komori/View/Weather/WeatherViewController.swift

```diff
@@ -42,9 +52,32 @@ class WeatherViewController: UIViewController {
         case .rainy:
             weatherImage.image = UIImage(named: "img_rainy")
             weatherImage.tintColor = .systemBlue
+        }
+    }
 
-        default:
-            return
+    private func showAlert(error: Error) {
+        let message = switch error {
+        case is YumemiWeatherError: "天気の取得に失敗しました。\n時間を置いて再度操作してください。"
+        default: "エラーが発生しました。\n時間を置いて再度操作してください。"
```

Comment:
```suggestion
        let message = switch error {
        case is YumemiWeatherError: "天気の取得に失敗しました。\n時間を置いて再度操作してください。"
        default: "エラーが発生しました。\n時間を置いて再度操作してください。"
```

エラー通知の粒度が大きいように感じるので、より精密に篩い分けてメッセージを変化させても良いかも知れません。

`YumemiWeatherError` は、現状では `invalidParameterError` と `unknownError` の２種類を返すため、そのどちらのエラーであるかを通知すると、ユーザーまたはそのサポート担当は問題を切り分けやすくなるかも知れないです。

また、`default` で想定されているとおり、それ以外のエラーが発生する可能性も考慮しておきたいところです。そしてこの場面での通知を `エラーが発生しました。` と括ると大きくなりすぎて（開発者は理解できても、ユーザーやそのサポート担当は、もしかすると `YumemiWeatherError` が示す状況でこのエラーが発生していると勘違いして）原因の特定が円滑に進まない可能性もあるかもしれないので、ここでは `YumemiWeatherError` 以外のエラー（たとえば「予期しないエラー」）であることを伝えておいても良いかもしれないです。

なお、`YumemiWeatherError` の詳細な篩い分けを行う場合、次のものが考えられそうです。

1. case で型キャストパターンを使い、列挙子パターンで篩い分ける
2. `LocalizedError` に対応させておき、case での列挙子パターンの篩い分けはせずに `localizedDescription` でエラーの詳細を取得してメッセージに埋め込む

Reply:
> 1. case で型キャストパターンを使い、列挙子パターンで篩い分ける

こちらの方法でエラーメッセージを修正しました！ 3568e93
また、アラート用のタイトル・メッセージの取得処理をextensionに切り出してみました。 bfdb46f

Reply:
![good-badge](https://img.shields.io/badge/review-good-blightgreen.svg)

`YumemiWeatherError` の列挙子に応じてメッセージが切り替わってとても良くなった印象です。もっと大きな視点では、天気予報のエラーと不明なエラーとで明確に切り分けられるタイトルになっているのも好印象です。


![next-badge](https://img.shields.io/badge/review-next-blueviolet.svg)

`extension` でどこまで切り出せるかを考えて最適化を図ってみると、より良いコードになりそうに思えました。


![imo-badge](https://img.shields.io/badge/review-imo-orange.svg)

エラー型はエラーの内容を表現できますけれど、それがタイトル付きのアラートで使われるかまでは想定できないように感じるため、エラー型で `alertTitle` や `alertMessage` プロパティーを備えるのは少し早そうです。これについては `extension` で処理を切り出す前の、アラートを表示する仕事を受け持つ `showAlert` が担当するのが妥当のように感じます。

エラーメッセージの取得に関しては、`alertMessage` プロパティーを `Error` プロトコルに備えることにすると、すべてのエラーを `Error` プロトコルが把握しないといけなくなり、荷が重そうに感じます。

各エラーの内容は、それを表現する各エラー型（今回であれば `YumemiWeatherError`）自身がいちばん把握しやすいと思うのと、Swift では「LocalizedError プロトコルに準拠させることで、そのインスタンスが示すエラーの内容をテキストで取得できる」という共通認識でコードが書ける作りになっているため、これを活用して `YumemiWeatherError` 自身にエラーメッセージを取得する機能を備えられます。

```swift
extension YumemiWeatherError : LocalizedError {
    
    public var errorDescription: String? {
        // ここで、列挙子ごとにメッセージを切り替える。
    }
}
```

このようにすることで、`YumemiWeatherError` 型の `error` が意味する内容（エラーメッセージ）はいつでも `error.localizedDescription` プロパティーを通して取得できます。それ以外の型で表現される `error` についても、必ず `error.localizedDescription` でそのテキスト表現にアクセスできるようになっている約束事のおかげで、詳細な篩い分けをしなくても、エラーメッセージの切り分けが可能になります。

極端な例を挙げると次のようになります。

```swift
private func showAlert(error: Error) {
    let alertController = UIAlertController(
        title: "エラー発生",
        message: error.localizedDescription,
        preferredStyle: .alert
    )

```

ただ、現在の `alertTitle` で実現されている「エラーの大まかな分類でタイトルを切り替える」のも効果的に感じるため、実際にコーディングする際には `showAlert` メソッド内でも大まかなエラーの切り分け（現在の `alertTitle` の実装のように）が必要になってきそうな予感がします。

```swift
private func showAlert(error: Error) {

    let title = switch error {
        case is YumemiWeatherError: "天気予報エラー"
        default: "不明なエラー"
    }

    let message = error.localizedDescription

```

Reply:
ありがとうございます。

>エラー型はエラーの内容を表現できますけれど、それがタイトル付きのアラートで使われるかまでは想定できないように感じるため、エラー型で alertTitle や alertMessage プロパティーを備えるのは少し早そうです。

アラートを表示する前提で考えてしまっていました...
Errorのextensionは廃止し、LocalizedErrorプロトコルに準拠させる方法で修正してみようと思います。

approveしていただいているので、一旦マージし、次のセッションにて修正します。